crates/autopilot/src/run_loop.rs

```diff
@@ -62,42 +72,77 @@ impl RunLoop {
         loop {
-            let init_block_timestamp = {
-                if let RunLoopMode::SyncToBlockchain = self.synchronization {
-                    let block = ethrpc::block_stream::next_block(self.eth.current_block()).await;
-                    if let Err(err) = self.solvable_orders_cache.update(block.number).await {
-                        tracing::error!(?err, "failed to build a new auction");
-                    }
-                    block.timestamp
-                } else {
-                    tokio::time::sleep(Duration::from_secs(1)).await;
-                    self.eth.current_block().borrow().timestamp
-                }
-            };
-
+            self.sleep_until_next_auction().await;
             if let Some(domain::AuctionWithId { id, auction }) = self.next_auction().await {
-                let current_block = self.eth.current_block().borrow().hash;
+                let current_block = *self.eth.current_block().borrow();
                 // Only run the solvers if the auction or block has changed.
                 let previous = last_auction.replace(auction.clone());
                 if previous.as_ref() != Some(&auction)
-                    || last_block.replace(current_block) != Some(current_block)
+                    || last_block.replace(current_block.hash) != Some(current_block.hash)
                 {
                     observe::log_auction_delta(id, &previous, &auction);
                     self.liveness.auction();
 
-                    self.single_run(id, &auction, init_block_timestamp)
+                    self.single_run(id, &auction, current_block.timestamp)
                         .instrument(tracing::info_span!("auction", id))
                         .await;
                 }
             };
         }
     }
 
+    /// Sleeps until the next auction is supposed to start and returns the most
+    /// recent block for that auction.
+    async fn sleep_until_next_auction(&self) {
+        match self.synchronization {
+            RunLoopMode::Unsynchronized => {
+                // Sleep a bit to avoid busy loops.
+                tokio::time::sleep(std::time::Duration::from_millis(1_000)).await;
+            }
+            RunLoopMode::SyncToBlockchain => {
+                let current_block = *self.eth.current_block().borrow();
+                let time_since_last_block = current_block.observed_at.elapsed();
+                let auction_block = if time_since_last_block > self.max_runloop_delay {
+                    tracing::warn!(
+                        missed_by = ?time_since_last_block - self.max_runloop_delay,
+                        "missed optimal auction start, wait for new block"
+                    );
+                    ethrpc::block_stream::next_block(self.eth.current_block()).await
+                } else {
+                    current_block
+                };
+
+                self.run_maintenance().await;
+
+                if let Err(err) = self
```

Comment:
Can move solvable orders cache also to maintenance (so it's treated the same way as event processing, which I believe at least logically is already the case)

Reply:
Overall I agree that conceptually this logic fits into the concept of `maintenance`.
However, this could cause issues when the `SolvableOrdersCache` logic runs before the events are indexed. Then we would for example still include orders that were settled in the previous block.
At the moment we don't have a nice way to control which `Maintaining` task depends on which other task.

In the most general case I guess we'd need to have a dependency graph of tasks and have `ServiceMaintenance` run the graph in topological order. But this seems very complex and we can probably get by with 2 adapters:
1. ParallelMaintenance
2. SerialMaintenance

With those our setup could look like this (which I already find too complex):
```
Serial(
    Parallel(
        Serial(SettlementEvents, OnSettlementEventUpdater),
        Parallel(AnyOtherEvents)
    ),
    SolvableOrdersCache
)
```

But since the task we try to solve should be relatively simple and this is getting quite complex I'm starting to lean towards a separate component like `AutopilotMaintenance` that takes all the individual `dyn Maintaining` and explicitly implements their dependencies. 🤔

Reply:
> However, this could cause issues when the SolvableOrdersCache logic runs before the events are indexed.

I think we could just simply create an autopilot domain maintenance class which itself implements `Maintaining` and contains an event updater and the solvable order cache and calls them in the order we require them to be called.

No need to abstract all possible dependency graphs into the higher level trait.

Reply:
Yeah, this is what I eluded to in the last paragraph and already started working on. Seems like the most reasonable solution at the moment.